### ISSUE TITLE
RFE: CGAL_DEPRECATED should be versionned

### ISSUE THREAD

**Issue by lrineau** [CGAL/cgal#5361]
### TL;DR

- We need a macro `CGAL_DEPRECATED` macro with a version, like in `CGAL_DEPRECATED_SINCE(5,3)`.
- `CGAL_DEPRECATED` without a message should itself be deprecated, in favor of `CGAL_DEPRECATED_MSG`.

### Rational

I am working with a team of developers on a project using CGAL, that assumes the use of CGAL-5.2 as the minimal version of CGAL (for the moment). I, personally, almost always use the `master` branch of CGAL.

The code uses [`write_ply`](https://doc.cgal.org/5.2/Surface_mesh/group__PkgSurface__mesh.html#ga77bbb79d449c981895eedb6c3c23bd14), but in `master` [that function has been deprecated](https://github.com/CGAL/cgal/blob/0419d3c42bf95c4761cc26a0a3bef6dc18f5ec57/Surface_mesh/include/CGAL/Surface_mesh/IO/PLY.h#L1133-L1142) (since yesterday and the merge of https://github.com/CGAL/cgal/pull/4255). So now I have a deprecation warning, and I cannot change the code base because other developers in the team maybe using a different version of CGAL. And I cannot use `#if` tricks with `CGAL_VERSION_NR` because there is no way to distinguish the version of CGAL just before the merge of #4255 and the version just after its merge.

That is why I think CGAL should offer to its users a way to setup a minimal version of CGAL to trigger the deprecation warnings. Something similar to what is offered by Qt5:

```
  QT_DEPRECATED_VERSION(major, minor) and QT_DEPRECATED_VERSION_X(major, minor, text)
  outputs a deprecation warning if QT_DEPRECATED_WARNINGS_SINCE is equal or greater
  than the version specified as major, minor. This makes it possible to deprecate a
  function without annoying a user who needs to stick at a specified minimum version
  and therefore can't use the new function.
```
See https://github.com/qt/qtbase/blob/13c460d0ff1a4eecfb7b1bc43a863783ed59a2bd/src/corelib/global/qglobal.h#L323-L381 for the details of the implementation.

## Proposal

The new CGAL macros should be named `CGAL_DEPRECATED_SINCE(x, y)` and `CGAL_DEPRECATED_SINCE_MSG(x, y, msg)` (or `CGAL_DEPRECATED_SINCE_X(x, y, msg)`, `_X` stands for "extended").

I also think that our guidelines should forbid the use of any macro but `CGAL_DEPRECATED_SINCE_MSG`... And if we agree on that, then the `CGAL_DEPRECATED_SINCE_MSG` and `CGAL_DEPRECATED_SINCE` could be merged into the simple name `CGAL_DEPRECATED_SINCE`.

(Note that Qt also has a versioned equivalent to our macro `CGAL_NO_DEPRECATION_CODE`. I also think it is a good idea.)

**Comment by mglisse** [comment]
> And I cannot use `#if` tricks with `CGAL_VERSION_NR` because there is no way to distinguish the version of CGAL just before the merge of #4255 and the version just after its merge.

You want to use one code for 5.2 (and 5.2.x-branch) and another one for master. I don't think it is very important to support non-current versions of master that do not have a release associated to them, so you wouldn't need to check exactly if #4255 was merged. Then the only point is to make sure that CGAL_VERSION_NR is increased right before/after a release. For instance gcc uses .0.0 version numbers for master, .1 for release branches, and .n.0 for official releases.

The deprecation warning is a special case where the code still works. But similar cases could exist for breaking changes, where you cannot just ignore the warning. In some cases, we could define a feature macro (like the C++ standard does). Or since we have a central repository, we could also have each merge increase the version number.

Documenting since when something is deprecated does seem useful, and versioned macros to say which deprecation warnings I want to see look like a good idea.
